ql/src/java/org/apache/hadoop/hive/ql/udf/UDFSubstr.java

```diff
@@ -64,19 +64,39 @@ public UDFSubstr() {
     r = new Text();
   }
 
-  public Text evaluate(Text t, IntWritable pos, IntWritable len) {
+  public Text evaluate(Text t, LongWritable pos, LongWritable len) {
+    if ((t == null) || (pos == null) || (len == null)) {
+      return null;
+    }
+
+    long longPos = pos.get();
+    long longLen = len.get();
+    // If an unsupported value is seen, we don't want to return a string
+    // that doesn't match what the user expects, so we return NULL (still
+    // unexpected, of course, but probably better than a bad string).
+    if (longPos > Integer.MAX_VALUE || longLen > Integer.MAX_VALUE) {
```

Comment:
Should negative overflow also be handled?
```
long l = Integer.MIN_VALUE - 10L;
int i = (int)l;
```
I think `i` is going to be a positive integer.